src/jupyterhub_config/custom_spawner.py

```diff
@@ -14,4 +16,43 @@ def start(self):
         variables, and sets the notebook directory before starting the server.
         """
 
-        return super().start()
\ No newline at end of file
+        username = self.user.name
+
+        # Ensure the system user exists
+        self._ensure_system_user(username, group='jupyterhub')
+
+        return super().start()
+
+    def _ensure_system_user(self, username: str, group: str = None):
+        """
+        Create a system user with the given username if it does not already exist.
+        Ensure the group exists before creating the user.
+        """
+        try:
+            # Check if user already exists
+            result = subprocess.run(['id', username], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            if result.returncode == 0:
+                self.log.info(f'User {username} already exists')
+                return
+
+            # Create the user
+            self.log.info(f'Creating system user: {username}')
+            useradd_cmd = ['sudo', 'useradd', '-r']
+
+            if group:
```

Comment:
What is the plan around adding all the users to the jupyterhub group?

Reply:
I don’t have any. We are not there yet. Maybe assign group by OAuth or some type of user DB.

Reply:
Ok, so why does the code add the users to the group then?

Reply:
think it's good to do so. Otherwise we end up with users with no group.

Reply:
Ok, so this is an add a group because users need a group type thing. But that also means that users will be able to read and write each other's files by default unless the chmod 700 them. Is that the desired behavior?

Reply:
That's how NERSC does. KBase group members can read each other's files.

Reply:
Ok, if that's intentional :+1: 

I presume we'd have to change this if non-kbase users started using this system, but that's a ways off